src/metabase/lib/fe_util.cljc

```diff
@@ -71,54 +87,177 @@
-    (expand-temporal-expression expression-clause)
-    expression-clause))
+(defn- case-or-if-expression?
+  [clause]
+  (and (vector? clause)
+       (boolean (#{:case :if} (first clause)))))
+
+(defn- flatten-case-or-if-expression
+  "case and if expressions expect a vector of pairs of if-then clause as
+   the first argument, but ExpressionParts can only represent a flat list of clauses.
+
+   This helper ungroups the pairs into a flat list of arguments for expression-parts."
+  [[op options clause-pairs fallback]]
+  (let [clauses (into [] cat clause-pairs)
+        clauses-with-fallback (if (nil? fallback)
+                                clauses
+                                (conj clauses fallback))]
+    (into [op options] clauses-with-fallback)))
+
+;; HACK: This is a hack to make sure that the display name of an unknown field
+;; is "Unknown Field" instead of it's id.
+(defn- set-display-name-for-unknown-field
+  [metadata]
+  (if (and (nil? (:id metadata)) (nil? (:lib/expression-name metadata)))
+    (assoc metadata :display-name (i18n/tru "Unknown Field"))
+    metadata))
 
 (defn- column-metadata-from-ref
   [query stage-number a-ref]
-  (lib.filter/add-column-operators
-   (lib.field/extend-column-metadata-from-ref
-    query stage-number
-    (lib.metadata.calculation/metadata query stage-number a-ref)
-    a-ref)))
-
-(mu/defn expression-parts :- ExpressionParts
-  "Return the parts of the filter clause `expression-clause` in query `query` at stage `stage-number`."
-  ([query expression-clause]
-   (expression-parts query -1 expression-clause))
+  (as->
+   (lib.metadata.calculation/metadata query stage-number a-ref) metadata
+    (lib.field/extend-column-metadata-from-ref query stage-number metadata a-ref)
+    (lib.filter/add-column-operators metadata)
+    (set-display-name-for-unknown-field metadata)))
+
+(defn- segment-metadata-from-ref
+  [query segment-ref]
+  (if-let [segment (lib.metadata/segment query (last segment-ref))]
+    segment
+    {:lib/type :metadata/segment
+     :id (last segment-ref)
+     :display-name (i18n/tru "Unknown Segment")}))
+
+(defn- metric-metadata-from-ref
+  [query metric-ref]
+  (if-let [metric (lib.metadata/metric query (last metric-ref))]
+    metric
+    {:lib/type :metadata/metric
+     :id (last metric-ref)
+     :display-name (i18n/tru "Unknown Metric")}))
+
+(defmulti expression-parts-method
+  "Builds the expression parts by dispatching on the type of the argument."
+  {:arglists '([query stage-number arg])}
+  (fn [_query _stage-number value]
+    (lib.dispatch/dispatch-value value))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod expression-parts-method :default
+  [query stage-number value]
+  (let [[op options & args] (cond-> value
+                              (expandable-temporal-expression? value) expand-temporal-expression
+                              (case-or-if-expression? value) flatten-case-or-if-expression)]
```

Comment:
Part of the power of having a multimethod is that these special cases can be handled at the top.

`(lib.dispatch/dispatch-value [:if ...])` is `:if`, so you can have a `(defmethod expression-parts-method :if ...)` that handles the case specifically. To reuse this `:default` case, you can either factor it out into a named function, or use the core function `(get-method some-multimethod dispatch-value)` like this:

```clojure
((get-method expression-parts-method :default) query stage-number adjusted-if-clause)
```

We can pair tomorrow to work through these refactors, if you like.